[CIR][CIRGen] Add dynamic builtin alloca intrinsics support

orbiri opened the pull request:
This patch adds the CIRGen for the following builtin functions:

- `alloca`;
- `_alloca`;
- `__builtin_alloca`;
- `__builtin_alloca_uninitialized`.

Missing support to add in the future:
- Non-default auto initialization setting. The default is to not initialize the allocated buffer, which is simpler to implement. This commit is leaving the skeleton to implement this feature following clang's codegen pattern.
- It may be possible that the frontend has set non-default address space for the alloca's return value. This is the case for OpenCL or AMDGPU codes for example. This is handled in clang codegen via address space cast, and is left for future implementation. This commit introduces a guard-rail around this behaviour.

----------------

P.S.:
I'm aware this is rather an "esoteric" feature to be supported and I'm sure there are more valuable features to implement before that! Implementing this feature was my method of learning more about the development flow for CIR - working along side clang's codegen, testing and debugging flow, etc.

orbiri commented:
Embarrassing formatting and lit-tests errors fixed and verified locally 🤦

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -642,6 +660,48 @@ RValue CIRGenFunction::buildBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+    // FIXME: It may make sense to allow AllocaOp of type `u8` to return a
+    // pointer of type `void *`. This will require a change to the allocaOp
+    // verifier.
+    auto Alloca = builder.create<mlir::cir::AllocaOp>(
```
CIRGenBuilder should have its own `createAlloca`, like it does for other operations. We usually do a best effort to add those (see both `CIRGenBuilder.h` and `CIRBaseBuilder.h`) as we go, and looks like this is a good moment to add this one (even though we missed some places in the past). CIRGen usually uses buildAlloca, but that expectes CIRGen specific stuff, we also need a createAlloca to abstract the actual materialization of `builder.create`.

orbiri replied:
I would do this happily! 
Can you help me understand the underlying philosophy/design behind it? If there's something I can read it would be even better than you reciting it for me 😇  
In addition, I'd be happy for a clarification on this comment:

> we also need a createAlloca to abstract the actual materialization of `builder.create`.

`CIRGenBuiltin.cpp` is full of direct calls to `builder.create<>()`. Is it just "legacy" code expecting to be properly wrapped over time? :)

orbiri replied:
And another note: seems like this isn't implemented in clang's derived `CGBuilderTy`, but only in the base `IRBuilderBase` class. I guess this is not applicable in MLIR since the operations are derived from the various dialects. This fact raises even more questions for me on the choice to hide the create materialization (as it isn't hidden for SCF ops I've sampled, for example 😇 ). I think I'd stop wondering about it and wait for your answer! 🙏

bcardosolopes replied:
> Can you help me understand the underlying philosophy/design behind it

It's easier to call into the helper function than having to figure out which version of Op::build to use most of the times. LLVM also has these convenient helpers and are a good source of inspiration.

> Is it just "legacy" code expecting to be properly wrapped over time?

Yes.

> And another note: seems like this isn't implemented in clang's derived CGBuilderTy, but only in the base IRBuilderBase class

We have things in both `CIRGenBuilder.h` and `CIRBaseBuilder.h` depending on whether the helpers are free of CIRGen specific params or not. Right now we have things in `CIRGenBuilder.h` that should be moved to  `CIRBaseBuilder.h`, we are currently moving those as we see need, but PRs to tackle that are also welcome.

bcardosolopes commented on clang/test/CIR/CodeGen/builtin-ms-alloca.c:
```diff
@@ -0,0 +1,23 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fms-extensions -emit-cir %s -o - | FileCheck %s --check-prefix=CIR
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fms-extensions -fclangir-enable -emit-llvm %s -o - | FileCheck %s -check-prefix=LLVM
```
Good job going the extra length to add `-fms-extension`, thanks!

bcardosolopes reviewed (changes requested):
Thanks for your first PR, looking forward to see more contributions :)

> I'm aware this is rather an "esoteric" feature to be supported and I'm sure there are more valuable features to implement before that! Implementing this feature was my method of learning more about the development flow for CIR - working along side clang's codegen, testing and debugging flow, etc.

No sweat, feel free to work on whatever makes you excited. Btw, these are actually important, since they show up in base system libraries, which are crucial to get CIR support bottom up in common software stacks.

orbiri commented:
Thanks for the review!! Will address the comments in the coming days.
New push is just rebasing on top of Nathan's rebase, didn't address any comment yet! :)

orbiri commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -642,6 +660,48 @@ RValue CIRGenFunction::buildBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+    // default (e.g. in C / C++ auto vars are in the generic address space). At
+    // the AST level this is handled within CreateTempAlloca et al., but for the
+    // builtin / dynamic alloca we have to handle it here.
+    // FIXME: introduce proper support for this feature like clang codegen has.
```
It seems like address spaces are NYI in CIR. I believe that adding proper support for address spaces in the CIR infrastructure may require deeper thought. I can attempt to tailor exactly what I need for this PR, but perhaps it is a bit out of scope.

Wdyt?

bcardosolopes replied:
What I'm suggesting here is that you add the skeletons to make this look like traditional LLVM codegen, it's fine you might not use this for now. You should also add `assert(!UnimplementedFeature::addressSpace());` here.

orbiri replied:
Since `getASTAllocaAddressSpace()` is already declared as an `UnimplementedFeature`, and since - not coincidentally - `createTempAlloca` relies on it staying unimplemented, I would want to suggest again to delay this to future PR. I will obviously add both asserts here to make sure this code is fixed once address spaces are supported.

Another option would be to implement it but to always return the default address space, and assert in this new implementation that addressSpace handling is not yet supported.

Do you have a preferred path forward? Or perhaps a third alternative?

orbiri replied:
Would adding the following to `TargetCIRGenInfo` make sense to you?
```
  /// Get the AST address space for alloca.
  virtual LangAS getASTAllocaAddressSpace() const {
    // Address spaces are not yet fully supported, but the usage of the default
    // alloca address space can be used for now only for comparison with the
    // default address space.
    assert(!UnimplementedFeature::addressSpace);
    return LangAS::Default;
  }
```

This seems reasonable, but has potential to blow up in someone's face if used improperly.

bcardosolopes replied:
> Since getASTAllocaAddressSpace() is already declared as an UnimplementedFeature, and since - not coincidentally - createTempAlloca relies on it staying unimplemented, I would want to suggest again to delay this to future PR. I will obviously add both asserts here to make sure this code is fixed once address spaces are supported.

I missed that, good point.

> Another option would be to implement it but to always return the default address space, and assert in this new implementation that addressSpace handling is not yet supported.

That'd be my ideal option, so we can probably diminish the amount of places we'll need to add such assert, luckily they will be all abstracted by `getASTAllocaAddressSpace`.

> Would adding the following to TargetCIRGenInfo make sense to you?

You should add the assert to `CIRGenTypeCache::getASTAllocaAddressSpace` (plz add that one for consistency). For  `TargetCIRGenInfo::getASTAllocaAddressSpace` you don't need the assert, because specific behavior needs to be added by targets anyways, and the default method in LLVM traditional codegen just simply returns `LangAS::Default` too.

bcardosolopes replied:
Great, just applied a review there too!

orbiri commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -642,6 +660,48 @@ RValue CIRGenFunction::buildBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+    // semantics of C/C++
+    // FIXME: It may make sense to allow AllocaOp of type `u8` to return a
+    // pointer of type `void *`. This will require a change to the allocaOp
+    // verifier.
```
Opened a tracking issue - #554

orbiri commented:
Addressed all comments, main changes:
- Addressed ASTAllocaAddressSpace support in a separate PR (#551 )
- Added `createAlloca` builder helpers in `CIRBaseBuilder.h`, and replaced usages with it.

orbiri commented:
@bcardosolopes do you think we can land this? :)
